src/order/interval.lean

```diff
@@ -0,0 +1,125 @@
+
+/-!
+# Order intervals
+
+This file defines intervals in an order as pairs where the second element is greater than the first.
+This is a prototype for interval arithmetic.
+-/
+
+open function order_dual set
+
+/-- The intervals in an order. -/
+@[ext] structure interval (α : Type*) [has_le α] extends α × α :=
+(fst_le_snd : fst ≤ snd)
+
+namespace interval
+variables {α β γ : Type*}
+
+section has_le
+variables [has_le α] {a b : interval α}
+
+/-- The injection from which we get the order. -/
+def to_dual_prod : interval α → αᵒᵈ × α := interval.to_prod
+
+lemma to_dual_prod_injective : injective (to_dual_prod : interval α → αᵒᵈ × α) :=
+λ a b, (ext_iff _ _).2
+
+instance [is_empty α] : is_empty (interval α) := ⟨λ a, is_empty_elim a.fst⟩
+instance [subsingleton α] : subsingleton (interval α) := to_dual_prod_injective.subsingleton
+
+instance : has_le (interval α) := ⟨λ a b, b.fst ≤ a.fst ∧ a.snd ≤ b.snd⟩
+
+lemma le_def : a ≤ b ↔ b.fst ≤ a.fst ∧ a.snd ≤ b.snd := iff.rfl
+
+/-- Turn an interval into an interval in the dual order. -/
+def dual : interval α ≃ interval αᵒᵈ :=
+{ to_fun := λ a, ⟨a.to_prod.swap, a.fst_le_snd⟩,
+  inv_fun := λ a, ⟨a.to_prod.swap, a.fst_le_snd⟩,
+  left_inv := λ a, ext _ _ $ prod.swap_swap _,
+  right_inv := λ a, ext _ _ $ prod.swap_swap _ }
+
+@[simp] lemma fst_dual (a : interval α) : a.dual.fst = to_dual a.snd := rfl
+@[simp] lemma snd_dual (a : interval α) : a.dual.snd = to_dual a.fst := rfl
+
+end has_le
+
+section preorder
+variables [preorder α] [preorder β] [preorder γ]
+
+instance : preorder (interval α) := preorder.lift interval.to_dual_prod
+
+/-- `{a}` as an interval. -/
+@[simps] def pure (a : α) : interval α := ⟨⟨a, a⟩, le_rfl⟩
+
+lemma pure_injective : injective (pure : α → interval α) := λ a b, congr_arg $ prod.fst ∘ to_prod
+
+@[simp] lemma dual_pure (a : α) : (pure a).dual = pure (to_dual a) := rfl
+
+instance [inhabited α] : inhabited (interval α) := ⟨pure default⟩
+instance : ∀ [nonempty α], nonempty (interval α) := nonempty.map pure
+instance [nontrivial α] : nontrivial (interval α) := pure_injective.nontrivial
+
+/-- Pushforward of intervals. -/
+@[simps] def map (f : α →o β) (a : interval α) : interval β :=
+⟨a.to_prod.map f f, f.mono a.fst_le_snd⟩
+
+@[simp] lemma map_pure (f : α →o β) (a : α) : (pure a).map f = pure (f a) := rfl
+@[simp] lemma map_map (g : β →o γ) (f : α →o β) (a : interval α) :
+  (a.map f).map g = a.map (g.comp f) := rfl
+
+@[simp] lemma dual_map (f : α →o β) (a : interval α) : (a.map f).dual = a.dual.map f.dual := rfl
+
+variables [bounded_order α]
+
+instance : order_top (interval α) :=
+{ top := ⟨⟨⊥, ⊤⟩, bot_le⟩,
+  le_top := λ a, ⟨bot_le, le_top⟩ }
+
+@[simp] lemma dual_top : (⊤ : interval α).dual = ⊤ := rfl
+
+end preorder
+
+section partial_order
+variables [partial_order α]
+
+instance : partial_order (interval α) := partial_order.lift _ to_dual_prod_injective
+
+/-- Consider an interval `[a, b]` as the set `[a, b]`. -/
+def to_set : interval α ↪o set α :=
```

Comment:
Is there anything that makes `Icc` a more natural choice than any other of the four interval constructors?

Reply:
(perhaps the answer is "it avoids difficulties with empty intervals"

Reply:
I don't think there's anything fundamental except that all references I looked at considered intervals to map to `Icc` and that `≤` is usually better behaved than `<` (because it corresponds to morphisms in the order considered as a category, while `<` doesn't have such a nice interpretation).

Reply:
Now that the empty interval is added by hand, it makes much more sense to use `Icc`.